docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md

```diff
@@ -0,0 +1,123 @@
+* Mozilla is moving towards rapid releases for mobile code.
+* We are willing to accept the overhead of upgrading application-services more frequently in application repositories.
+
+## Considered Options
+
+* **(A) Keep our current release process**
+* **(B) Move to a rapid release process where changes in app-services are frequently merged into consumer applications**
+    * Phase 1: update our branching model and release process (Jan 2023 - Feb 2023)
+        * Update our branches to match how firefox-android works: a `main` branch that is used for the current nightly builds and a `releases_v[N]` for each previous release.
+        * Automate a nightly release of our `main` branch
+          * See the discussion at the bottom of this document for how this could work
+          * Note: regardless of which option we choose, we will no longer be using semver-style versions and also not following semver semantics.  We will be releasing breaking changes in a nightly cycle without bumping the major version.
+        * Manually handle patch releases from the `release_v*` branches
+          * Update our version number to `[firefox-version].[N+1]`
+          * Create a GitHub release from the branch
+          * We should avoid having unreleased code in these branches.  Releases should be been soon after new code is committed to those branches.
+          * Code should always be uplifted into these branches.  Before code is merged into the beta branch it should have already been merged into `main` and released as a nightly.  Before code is merged into `release_v[N]` branch, it should have merged and released in the `release_v[N-1]` branch.
+        * Continue to maintain our CHANGELOG, but with new versioning.  In main, keep a rolling log of updates with the header `[firefox-version].0`
+        * This replaces our current release process.
+        * To push changes to consumers:
+           * Create a PR in their repository that bumps the application-services version to one of these releases.
+           * Don't merge any of these PRs unless all pass CI and are approved.
+           * Once the PRs are approved, then merge them all at once.
+        * We will implement a script which will be manually run on the monthly Firefox merge date. This script will:
+          * Create a `release_v[N]` branch of of main
+          * Create a GitHub release for that branch
+          * Bump the major version number in main and start a new changelog header.
+    * Phase 2: Automate merges of `application-services` into consumer applications (March 2023 - May 2023)
+      * Every night, for each consumer application (including all mobile apps and Desktop):
+          * Create a PR to take the current application-services code in `main`.
+          * Set up an auto-merge system to automatically merge the PRs.  As with phase 1, no PR should be merged until all PRs pass CI and then all PRs should be merged at once.
```

Comment:
> Who's going to be in charge of watching these PRs across every repo and manually merging them all at the same time? 

I was envisioning some automation that did that, but that might be magical thinking since I don't know how it would be implemented.

The alternative would be to auto-merge the PR for each repo if it passes the CI for that repo, without paying attention to the other CI.  That has a very clear implementation path, just extend the existing integrations to auto-merge.

The issue is that maybe the CI passes for one repo, but not another, and they get out of sync.  In general, that's not so bad and usually the worst case scenario is that you just have to revert the change and everything is fixed in the next nightly.  However, sometimes that won't be possible, for example if a change contains a DB migration.

I really like your suggestion of auto-merging without checking the other repos, but having a PR flag to prevent that behavior and require manual merging for the special cases where we can't simply revert the code.  What do others think about this?

Reply:
As mentioned above, we can couple auto-merges with a strict backout policy for breaking changes to ensure that things don't remain out of sync for long when they do break.